src/xdp/trn_transit_xdp_hardware_offload.c

```diff
@@ -0,0 +1,548 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <linux/bpf.h>
+#include <linux/if_arp.h>
+#include <linux/if_ether.h>
+#include <linux/if_packet.h>
+#include <linux/if_vlan.h>
+#include <linux/in.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <linux/pkt_cls.h>
+#include <linux/socket.h>
+#include <linux/tcp.h>
+#include <linux/types.h>
+#include <linux/udp.h>
+#include <stddef.h>
+#include <string.h>
+#include "extern/bpf_endian.h"
+#include "extern/bpf_helpers.h"
+#include "trn_datamodel.h"
+#include "trn_kern.h"
+
+int _version SEC("version") = 1;
+
+struct bpf_map_def SEC("maps") networks_offload_map = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct network_key_t),
+	.value_size = sizeof(struct network_offload_t),
+	.max_entries = 1000001,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(networks_offload_map, struct network_key_t, struct network_offload_t);
+
+struct bpf_map_def SEC("maps") vpc_offload_map = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct vpc_key_t),
+	.value_size = sizeof(struct vpc_offload_t),
+	.max_entries = 1000001,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(vpc_offload_map, struct vpc_key_t, struct vpc_offload_t);
+
+struct bpf_map_def SEC("maps") endpoints_offload_map = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct endpoint_key_t),
+	.value_size = sizeof(struct endpoint_offload_t),
+	.max_entries = 1000001,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(endpoints_offload_map, struct endpoint_key_t, struct endpoint_offload_t);
+
+struct bpf_map_def SEC("maps") interface_config_offload_map = {
+	.type = BPF_MAP_TYPE_ARRAY,
+	.key_size = sizeof(int),
+	.value_size = sizeof(struct tunnel_iface_t),
+	.max_entries = 1,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(interface_config_offload_map, int, struct tunnel_iface_t);
+
+static __inline int trn_rewrite_remote_mac(struct transit_packet *pkt)
+{
+	/* The TTL must have been decremented before this step, Drop the
+	packet if TTL is zero */
+	if (!pkt->ip->ttl)
+		return XDP_DROP;
+
+	struct endpoint_offload_t *remote_ep;
+	struct endpoint_key_t epkey;
+	epkey.tunip[0] = 0;
+	epkey.tunip[1] = 0;
+	epkey.tunip[2] = pkt->ip->daddr;
+	/* Get the remote_mac address based on the value of the outer dest IP */
+	remote_ep = bpf_map_lookup_elem(&endpoints_offload_map, &epkey);
+
+	if (!remote_ep) {
+		return XDP_DROP;
+	}
+
+	trn_set_src_mac(pkt->data, pkt->eth->h_dest);
+	trn_set_dst_mac(pkt->data, remote_ep->mac);
+
+	if (pkt->ip->tos & IPTOS_MINCOST) {
```

Comment:
I'm just leaving a comment for myself to follow-up. I need to see if this code (QoS) would be applicable here for offloaded case. No action needed yet.